### PR TITLE
 policy: add policy/mapstate/stage script command 

### PR DESCRIPTION
**Pull request by squeed**
This adds a command that takes an endpoint and a set of proposed
policies, and shows the effect this would have on the policy's mapstate.

It uses the PolicyRepository.Snapshot() functionality, which can freeze
the state of the PolicyRepository and the set of identities. It then
simulates ingesting the set of policies. It then displays the
resulting MapState (or the difference, as desired).

This required duplicating logic from a few places in the codebase. In
particular, copying toService extraction and CIDR identity generation
were both necessary.

```release-note
Added the `policy/mapstate/stage` command to `cilium-dbg shell`. It can show the difference in policy map entries that would be created by a proposed set of policies.
```

### CODE REVIEW

File: pkg/policy/commands/cell.go [hunk 1]
```diff
@@ -6,31 +6,55 @@ package commands
...
+	Repository policy.PolicyRepository
+	DB         *statedb.DB
+	Services   statedb.Table[*loadbalancer.Service]
+	Backends   statedb.Table[*loadbalancer.Backend]
+
+	ClusterMeshPolicyConfig cmtypes.PolicyConfig
+	Config                  *option.DaemonConfig
 }
 
+// Cell provides the PolicyRepository and PolicyUpdater.
```
Comment (bimmlerd) [inline pkg/policy/commands/cell.go]: nit: probably outdated comment?

File: pkg/policy/commands/cell.go [hunk 2]
```diff
@@ -39,17 +63,42 @@ func NewPolicyCommands(params CmdParams) hive.ScriptCmdsOut {
...
+	return filterPrefix(epNames, cur)
+}
 
-		return filterPrefix(epNames, cur)
+func autocompleteFilenameImpl(state *script.State, cur string) []string {
+	entries, err := os.ReadDir(state.Path(cur))
+	dir := cur
+	file := ""
+	if err != nil {
+		dir, file = path.Split(state.Path(cur))
```
Comment (bimmlerd) [inline pkg/policy/commands/cell.go]: this should use `path/filepath` instead: https://pkg.go.dev/path/filepath

File: pkg/policy/commands/cell.go [hunk 3]
```diff
@@ -39,17 +63,42 @@ func NewPolicyCommands(params CmdParams) hive.ScriptCmdsOut {
...
+			epNames = append(epNames, fmt.Sprintf("%s/%s", ep.K8sNamespace, ep.K8sPodName))
 		}
+	}
+
+	return filterPrefix(epNames, cur)
+}
 
-		return filterPrefix(epNames, cur)
+func autocompleteFilenameImpl(state *script.State, cur string) []string {
+	entries, err := os.ReadDir(state.Path(cur))
```
Comment (bimmlerd) [inline pkg/policy/commands/cell.go:85]: Could probably do this nicer with https://pkg.go.dev/path/filepath#WalkDir

File: pkg/policy/commands/cell.go [hunk 4]
```diff
@@ -39,17 +63,42 @@ func NewPolicyCommands(params CmdParams) hive.ScriptCmdsOut {
...
+		dir, file = path.Split(state.Path(cur))
+		entries, err = os.ReadDir(dir)
+		if err != nil {
+			return nil
+		}
+	}
+
+	var suggestions []string
+	for _, entry := range entries {
+		if file == "" || strings.HasPrefix(entry.Name(), file) {
```
Comment (bimmlerd) [inline pkg/policy/commands/cell.go]: nit: emptyness check redundant IMO, this isn't some hot function where it matters and it makes the code less readable

File: pkg/policy/commands/mapstate_diff.go [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+	policy.FallbackRedirectID: math.MaxUint16,
+}
+
+func msStageCmd(params CmdParams) script.Cmd {
+	return script.Command(
+		script.CmdUsage{
+			Summary: "Display resulting policy map for a proposed policy change.",
+			Flags: func(fs *pflag.FlagSet) {
+				fs.StringP("output", "o", "table", "Format to write in (table, json)")
+				fs.BoolP("diff", "d", false, "Display the difference between existing policy.")
```
Comment (bimmlerd) [inline pkg/policy/commands/mapstate_diff.go]: nit: maybe better `Display the different to existing policy`

File: pkg/policy/repository.go [hunk 6]
```diff
@@ -745,3 +745,41 @@ func RepositoryScriptCmds(p *Repository) map[string]script.Cmd {
...
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+
+	out := NewPolicyRepository(
+		logger,
+		p.selectorCache.GetIdentities(),
+		p.certManager,
+		p.l7RulesTranslator,
+		identitymanager.NewIDManager(logger),
+		p.metricsManager,
```
Comment (bimmlerd) [inline pkg/policy/repository.go]: Should this be a mock metrics manager? I think having this be the live one means that we add to metrics for simulated policy ingestion, which is probably incorrect.

I guess in general, did you verify that these other shared components don't have undesired side effects when being called by simulation?

File: pkg/policy/commands/mapstate_diff.go [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+				}
+
+				diff, err := state.Flags.GetBool("diff")
+				if err != nil {
+					return "", "", err
+				}
+
+				cmd, err := initialize(params, state)
+				if err != nil {
+					return "", "", nil
```
Comment (bimmlerd) [inline pkg/policy/commands/mapstate_diff.go]: nit: did you mean `return "", "", err`?

File: pkg/policy/commands/mapstate_diff.go [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+				if err != nil {
+					return "", "", fmt.Errorf("failed to apply new policy: %w", err)
+				}
+
+				entriesAfter, err = cmd.getEPEntries()
+				if err != nil {
+					return "", "", fmt.Errorf("failed to compute endpoint's policy map: %w", err)
+				}
+
+				res := toResult(entriesBefore, entriesAfter)
```
Comment (bimmlerd) [inline pkg/policy/commands/mapstate_diff.go:129]: I think res has all entries as "net new" entries if `--diff` is not set (which implies `entriesBefore` is empty). 

The `table` output distinguishes between being in diff mode or not, and if not just prints the mapstate, but the `json` output prints both the mapstate and each entry as a diff addition from what I can tell?

File: pkg/policy/commands/mapstate_diff.go [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+
+// getExistingEntries evaluates the policy state as-is for the given endpoint,
+func (s *stageCmd) getEPEntries() (map[policytypes.Key]entryOut, error) {
+	sp, err := s.pr.ResolvePolicy(s.epID)
+	if err != nil {
+		return nil, err
+	}
+	sp.Detach()
+
+	epp := sp.DistillPolicy(s.log, s.ep, dummyRedirects)
```
Comment (bimmlerd) [inline pkg/policy/commands/mapstate_diff.go]: AI raised that if you `DistillPolicy` after having detached the selector policy, it will cause endpoint regeneration. The fix seems to be to only detach the policy after distillation?

I tried to verify this and I think the issue raised is right:

`L4Policy.detach()` sets `l4.users = nil`, but `DistillPolicy` calls `insertUser` which contains
```
	if l4.users != nil {
		l4.users[user] = struct{}{}
	} else {
		go user.PolicyOwner.RegenerateIfAlive(&regeneration.ExternalRegenerationMetadata{
			Reason:            regeneration.ReasonSelectorPolicyStale,
			Message:           "selector policy has changed because of another endpoint with the same identity",
			RegenerationLevel: regeneration.RegenerateWithoutDatapath,
		})
	}
```
which means we call `RegenerateIfAlive`.

**Review by bimmlerd — CHANGES_REQUESTED** [review]
general direction LGTM, but found some issues

File: pkg/policy/repository.go [hunk 10]
```diff
@@ -745,3 +745,41 @@ func RepositoryScriptCmds(p *Repository) map[string]script.Cmd {
...
+	)
+	out.namedPortsGetter = p.namedPortsGetter
+
+	// Insert all rules in to the new policy repository.
+	//
+	// These need to be inserted as if they were coming externally, so selectors
+	// will be allocated in the new selectorcache.
+	for k, r := range p.rules {
+		newRule := out.newRule(r.PolicyEntry, k)
+		out.insert(newRule)
```
Comment (fristonio) [inline pkg/policy/repository.go:781]: Should we manually construct internal fields instead of using `Repository.insert()` so we don't pollute the metrics?

File: pkg/policy/k8s/service.go [hunk 11]
```diff
@@ -256,6 +257,21 @@ func (p *policyWatcher) resolveToServices(key resource.Key, cnp *types.SlimCNP)
...
 
+// ResolveToServices translates all ToServices rules found in the provided
+// CNP to CIDR rules. Mutates the supplied CNP. Used by other policy consumers.
+func ResolveToServices(txn statedb.ReadTxn, services statedb.Table[*loadbalancer.Service], backends statedb.Table[*loadbalancer.Backend], cnp *v2.CiliumNetworkPolicy) {
+	for svc := range services.All(txn) {
+		svcEndpoints := newServiceEndpoints(svc, txn, backends)
+
+		// This extracts the selected service endpoints from the rule
+		// and translates it to a ToCIDRSet/ToEndpoints
+		numMatches := svcEndpoints.processRule(cnp.Spec)
```
Comment (fristonio) [inline pkg/policy/k8s/service.go]: Return value(`numMatches`) seems unused, maybe use blank identifier(`_`)?

File: pkg/policy/commands/mapstate_diff.go [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+	MapState []entryOut  `json:"mapState"`
+	Diff     []diffEntry `json:"diff"`
+}
+
+type diffEntry struct {
+	Deleted *entryOut `json:"deleted,omitempty"`
+	Added   *entryOut `json:"added,omitempty"`
+}
+
+func initialize(params CmdParams, state *script.State) (*stageCmd, error) {
```
Comment (fristonio) [inline pkg/policy/commands/mapstate_diff.go]: Maybe rename to have explicit context for `stageCmd` to avoid ambiguity.

```suggestion
func initializeStageCmd(params CmdParams, state *script.State) (*stageCmd, error) {
```

File: pkg/policy/commands/mapstate_diff.go [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+	pr := params.Repository.(*policy.Repository)
+	if pr == nil {
+		return nil, fmt.Errorf("BUG: could not cast policy repository")
+	}
+	// Take a snapshot of the repository so we can make changes
+	s.pr = pr.Snapshot(s.log)
+
+	// add this endpoint to the subject selector cache
+	wg := sync.WaitGroup{}
+	s.pr.GetSubjectSelectorCache().UpdateIdentities(identity.IdentityMap{s.epID.ID: s.epID.LabelArray}, nil, &wg)
```
Comment (fristonio) [inline pkg/policy/commands/mapstate_diff.go:226]: While probably not relevant for policy repository snapshot we should `Wait` on wg for consistency?

File: pkg/policy/commands/mapstate_diff.go [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+		applied := false
+		for {
+			u := unstructured.Unstructured{}
+			if err := decoder.Decode(&u); err != nil {
+				if errors.Is(err, io.EOF) {
+					break
+				}
+				return fmt.Errorf("failed to parse policy: %w", err)
+			}
+			if err := s.applyPolicy(&u); err != nil {
```
Comment (fristonio) [inline pkg/policy/commands/mapstate_diff.go:251]: I think it would be helpeful to keep the parse and apply steps in separate loops. This will allow us to not fail in half built repository state.

File: pkg/policy/selectorcache.go [hunk 15]
```diff
@@ -874,3 +874,18 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted identity.IdentityMap, w
 	}
 	return mutated
 }
+
+// GetIdentities captures the current set of identities.
+// Used for debugging, staging, and testing.
+func (sc *SelectorCache) GetIdentities() identity.IdentityMap {
```
Comment (fristonio) [inline pkg/policy/selectorcache.go]: Wondering if we actually need to clone the `idCache` map given its already owned by the script command. This seems like an expensive operation and given the original id cache is already keyed on NumericIdentity we can maybe export the type and reuse it?

File: pkg/policy/repository.go (hunk 6 above)
Comment (squeed) [reply pkg/policy/repository.go]: Good point about the metrics manager. I was trying to avoid having mocks in non-test code but that's not actually the right idea.

I did check that the other shared components are "read-only".

The only one that is borderline is the certificate manager. It is read-only, but it does have k8s secret access and may change contracts on us in the future. It will be irrelevant once the transition to SDS is complete. I can require that SDS is enabled to use this command.

File: pkg/policy/commands/cell.go (hunk 1 above)
Comment (squeed) [reply pkg/policy/commands/cell.go]: Huh, must have been a mis-paste

File: pkg/policy/selectorcache.go (hunk 15 above)
Comment (squeed) [reply pkg/policy/selectorcache.go]: It doesn't seem _too_ expensive to me. We're not allocating any identities, just shuffling some pointers in to a map. This is a CLI tool for debugging, so I'm not super concerned about short-lived memory usage.

If we really wanted to save an alloc, we could return the dumped identities when snapshotting.

Comment (squeed) [reply pkg/policy/selectorcache.go]: In particular, I'd rather not make the selectorcache internals exported.

File: pkg/policy/commands/cell.go (hunk 3 above)
Comment (squeed) [reply pkg/policy/commands/cell.go:85]: Don't blame me, it's just copy-pasted from https://github.com/cilium/hive/blob/main/script/cmds.go#L98 :-)

File: pkg/policy/commands/mapstate_diff.go (hunk 7 above)
Comment (squeed) [reply pkg/policy/commands/mapstate_diff.go]: Sure did.

File: pkg/policy/commands/mapstate_diff.go (hunk 8 above)
Comment (squeed) [reply pkg/policy/commands/mapstate_diff.go:129]: The JSON output returns both the final mapstate as well as, (now) optionally, the diff. The Table output just shows the result or diff, since it's for human consumption. Seem reasonable?

File: pkg/policy/commands/mapstate_diff.go (hunk 14 above)
Comment (squeed) [reply pkg/policy/commands/mapstate_diff.go:251]: This is the staged repository anyways; it is discarded after the command is complete. So, success or failure, it all gets flushed.

File: pkg/policy/commands/mapstate_diff.go (hunk 9 above)
Comment (squeed) [reply pkg/policy/commands/mapstate_diff.go]: Good catch. Thanks, bot! I've wrapped the endpoint in a shim that blocks all side-effecting calls.

File: pkg/policy/repository.go (hunk 10 above)
Comment (squeed) [reply pkg/policy/repository.go:781]: I wrapped metrics in a nil check, this should be better now.

File: pkg/policy/selectorcache.go (hunk 15 above)
Comment (squeed) [reply pkg/policy/selectorcache.go]: I did remove the second call to GetIdentities though.

File: pkg/policy/commands/cell.go (hunk 3 above)
Comment (squeed) [reply pkg/policy/commands/cell.go:85]: actually, WalkDir isn't the right choice here, we don't want to recurse.

File: pkg/policy/repository.go (hunk 6 above)
Comment (squeed) [reply pkg/policy/repository.go]: I wound up mocking the certificate manager, just for safety's sake.

File: pkg/policy/commands/cell.go (hunk 3 above)
Comment (bimmlerd) [reply pkg/policy/commands/cell.go:85]: I don't think it's a bad choice, your `WalkFunc` can return https://pkg.go.dev/path/filepath#SkipDir when encountering a dir to not recurse into it. But I'm also not opinionated enough to say that it _must_ be `WalkDir`.

File: pkg/policy/commands/mapstate_diff.go (hunk 8 above)
Comment (bimmlerd) [reply pkg/policy/commands/mapstate_diff.go:129]: Ah, I missed the `before` nil check in `toResult` last time around, nevermind.

**Review by bimmlerd — APPROVED** [review]
LGTM

**Comment by squeed** [comment]
Not sure why the "build commits" step is failing, something weird about an empty go file on generate? Doesn't seem related to this PR.

File: pkg/policy/repository_test.go [hunk 16]
```diff
@@ -1465,3 +1465,67 @@ func TestReplaceByResource(t *testing.T) {
 	assert.True(t, sc.selectors.Empty())
 	assert.Equal(t, 2, oldRuleCnt)
 }
+
+func TestClone(t *testing.T) {
```
Comment (derailed) [inline pkg/policy/repository_test.go]: nit: TestSnapshot?

**Review by derailed — COMMENTED** [review]
@squeed Nice work!

File: pkg/policy/repository_test.go (hunk 16 above)
Comment (squeed) [reply pkg/policy/repository_test.go]: Yup.

File: pkg/policy/commands/mapstate_diff.go (hunk 14 above)
Comment (fristonio) [reply pkg/policy/commands/mapstate_diff.go:251]: Yeah, given this is a debug command I am not too concerned either, but wanted to avoid any unnecessery ops given this can be used on a live cluster. Resolving.

**Review by fristonio — APPROVED** [review]
🚀

**Comment by joestringer** [comment]
@squeed is this meant to refer to `cilium-dbg shell`?

```
# cilium-dbg script
Error: unknown command "script" for "cilium-dbg"
Run 'cilium-dbg --help' for usage.
unknown command "script" for "cilium-dbg"
```
